A basic REPL

steveklabnik opened the pull request:
This isn't perfect, but it's time to get some eyes on it at least.

This work not only contains a *very* basic REPL, but also a refactoring of the `init` functions used to hook subcommands into Humility. As we continue to develop the repl, the uniformity in these signatures is important, as we talked about in Emeryville.

I *also* included a couple of smaller naming changes that I think are illustrative, but I'm not wedded to them.

By "basic" repl I mean:

* you can type any command without 'humility' in front, and it will work
* I suggest setting `HUMILITY_ARCHIVE` before running `humility repl` so that you don't have to type `-a <path>` in front of every single line on the command (this is a thing I'd like to improve in the near future)
* there's one built-in command: `history`. This will show the commands you've run since the session started. 
* i have not yet integrated `rustyline`, though i hear it's easy. This means no nice keyboard shortcuts just yet. This is my number one priority after this work gets signed off on.

Happy for any and all feedback.

steveklabnik commented:
> i have not yet integrated rustyline,

https://github.com/oxidecomputer/crucible/issues/258

I don't know if illumos hosts are considered first class, but this would be a concern moving forward there.

rmustacc commented:
While many of us who are working on bringup and remote use are relying on humility on illumos and would likely take advantage of a repl when available, as long as we could use a non-repl based humility in the interim, I personally wouldn't block integration of rustyline on that getting fixed (we can do that in parallel).

steveklabnik commented:
Cool, yeah to be clear there is zero intention to remove the current, non-interactive CLI aspects of Humility. This is purely an opt-in kind of thing when you know you're going to be doing a lot of work in a row, and a REPL would be helpful.

steveklabnik commented:
Leaving this here for later: nushell uses https://github.com/nushell/reedline which may also be a good candidate. I haven't evaluated it other than "I use nushell every day."

leftwo commented:
> Leaving this here for later: nushell uses https://github.com/nushell/reedline which may also be a good candidate. I haven't evaluated it other than "I use nushell every day."

I just tried this and it works for me using the reedline sample program on my illumos system +tmux+mosh, as well as on my mac both inside and outside tmux.

steveklabnik commented:
I've pushed up two more commits: the first switches to `reedline`, and the second implements history via it, rather than my good old array of strings.

bcantrill reviewed (changes requested):
This is a good start, but there are definitely some structural things I'd like to consider here -- as well as some mechanical things like the need for documentation.  Also, just from running, I notice that typing `help` at the repl exits, which I don't think is the intent?

steveklabnik commented on humility-cmd/src/lib.rs:
```diff
@@ -88,22 +45,19 @@ pub enum Validate {
-        archive: Archive,
-        run: fn(&mut HubrisArchive, &Args, &[String]) -> Result<()>,
-    },
+pub struct AttachementMetadata {
```
Gosh darn it I checked this like five times I thought 🤣

steveklabnik commented on humility-cmd/src/lib.rs:
```diff
@@ -11,56 +11,13 @@ pub mod reflect;
 #[allow(dead_code)]
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
-pub enum Archive {
+pub enum ArchiveRequired {
```
I am not sure either! The main reason I was trying to find a new name here is that we also have `HubrisArchive`, and so when reading code using `Archive` I was always like "why is there an archive here?" when this is really answering the question about if we require an archive or not. I do think the name is a bit wonky, so I was trying out the McDonald's principle, haha. If there's no better name I'm happy to just revert it too.

steveklabnik commented on src/main.rs:
```diff
@@ -49,59 +38,44 @@ fn main() {
+    I: IntoIterator<Item = T>,
+    T: Into<OsString> + Clone,
+{
+    /*
```
haha, yeah, I don't know how this happened in this file; I will fix this for sure. Nice catch.

steveklabnik commented on src/cmd.rs:
```diff
@@ -16,21 +15,26 @@ use std::collections::HashMap;
+    // add in the repl
+    dcmds.push(CommandDescription {
+        init: repl::init,
+        docmsg: "For additional documentation, run \"humility doc repl\"",
```
If I could put it in another crate, I absolutely would, it's just that then it becomes a circular crate dependency, and so doesn't work :/

steveklabnik commented:
> Also, just from running, I notice that typing help at the repl exits, which I don't think is the intent?

Absolutely not, yep. Will investigate.

> Also, continuing to play around, it appears that running any command from the REPL that requires an archive fails:

Yes, this is what I mentioned above with the "I suggest setting HUMILITY_ARCHIVE before running humility repl"; this is my top thing to fix next; this code does not persist the "top-level" flags between invocations, so you have to type `-a ~/hubris/target/demo-stm32h753-nucleo/dist/build-demo-stm32h753-nucleo.zip tasks` in the repl, which sucks. Wanted to make sure this was all in the right direction before I started improving usability :)

Thanks as always for the review!

steveklabnik commented:
@bcantrill this is ready for a re-review. I have addressed various things in various ways:

> it appears that running any command from the REPL that requires an archive fails:

Turns out it was quite easy to persist these, so rather than doing it in a follow-up, I did it here.

> Right now, [`humility doc repl`] doesn't work:

I've hacked this in as we discussed in the huddle yesterday. For those who weren't there, the core issue here is that processing CLI options yields back a series of Commands. The repl wants to re-use this infrastructure, but that then leads to circular package dependencies. So instead we have to make it inside of the main package, and then hack in the documentation stuff. Oh well.

> AttachementMetadata is misspelled

I have backed these changes out. I wasn't so sure on them myself, and after sitting with the diff for a while, think it was a bad idea.

> you are accidentally reintroducing the comment style that I purged at your (implicit) urging. ;)

Addressed :)

steveklabnik commented on cmd/doc/build.rs:
```diff
@@ -46,6 +52,66 @@ fn cmd_docs(lookup: &str) -> Option<&'static str> {{
+Finally, as you can see, `quit` will quit the repl. There is also a `history`
+command, which will show you recent commands you've put into the prompt.
+
+The repl is still very early days! We'll be adding more features in the future.
```
I am happy to improve/add/whatever to these docs, but I figured I should have *something* there and we can adjust, at least this is enough to see how it looks in the file.

bcantrill commented:
I notice that running `help` or any of its variants (e.g., `tasks --help`) results in the REPL exiting.  I think we want to get this fixed, as I have to imagine that it will be a not infrequent thing to do. ;)  Note that we also need to get a help string for `repl` (running `humility --help` yields no description for `repl`).  And while I think this can be deferred until it's integrated, I think we are definitely going to want to get tab completion working, as that will also be a very popular reason to use the REPL. ;)

steveklabnik commented:
Okay, after trying like three different ways of accomplishing all of this, I think we're good to go. In the end, the path forward I went was to still make a `humility-cmd-repl` that's an empty crate; by naming the module with the same pattern, everything else just kinda shakes out properly, which is really nice.

steveklabnik commented on README.md:
```diff
@@ -1216,7 +1217,10 @@ VSC8552; this is indicated with `--` in the relevant table positions.
+This command launches OpenOCD based on the config file in a build archive,
+which then allows one to connect with either GDB or directly via telnet.
+If the intention is to only run GDB, note that `humility gdb --run-openocd`
+will both run OpenOCD and run a foreground GDB that is connected to it.
```
Fun fact: this change isn't due to me at all, but when I ran `cargo xtask readme`, it was generated. Guess that hadn't been done in a while.

bcantrill commented on src/cmd_repl.rs:
```diff
@@ -0,0 +1,142 @@
+
+    if let Err(e) = cmd::subcommand(context, &commands) {
+        Ok(format!(
+            "I'm sorry, Dave. I'm afraid I can't understand that: '{e}'",
```
This is cute,  but ultimately distracting. For example:

```console
humility> test
I'm sorry, Dave. I'm afraid I can't understand that: 'does not appear to be a test archive'
```
I think we probably should have something more pedestrian here that more closely approximates what one see when running the command without the REPL (e.g., `humility; {cmdname} failed: {errmsg}`)

steveklabnik replied:
I've pushed a change that simply shows the error, without the cutesy message. Output looks like this:

```
humility> tasks -f
error: Found argument '-f' which wasn't expected, or isn't valid in this context

        If you tried to supply `-f` as a value rather than a flag, use `-- -f`

USAGE:
    humility tasks [OPTIONS] [TASK]

For more information try --help
```

Given that it already mentions the command name in this output, I wonder if this is acceptable. If you want me to have it in the output anyway, I can do that, it'll just require moving some more code around (the command name isn't parsed out until further down the call stack).

steveklabnik replied:
(And yeah I did think it was cutesy but not really useful for the final form, so might as well get rid of this here :) )

bcantrill replied:
Looks great -- thank you!

bcantrill reviewed (commented):
I think this is really close! I don't want to further delay this, but I *do* think that there are some things that this is going to need to start to really be useful:

1. Failures currently bomb out entirely. For some commands, it is *really* common to have mistaken commands in part to figure out what it wants -- so bombing out is going to make it tough.

2. I'm not sure if it's something strange with my terminal, but when I edit a previous command  and delete characters, they do not appear deleted -- but they are in fact deleted.  (e.g., run `tasks -v`, then up arrow, then backspace, then backspace, `tasks -v` will still appear even though `tasks` will be run).

3. There are a couple of commands that will run in a loop (e.g. `sensors -s -t speed`); when these commands are ^C'd, the entire REPL bombs out.

4. Hitting enter on a blank line probably shouldn't be an error.

5. You can run `repl` from the REPL -- but no idea what this means.

I think these issues are probably worth fixing before this integrates -- but I'm personally really looking forward to using it!

steveklabnik commented on cmd/apptable/src/lib.rs:
```diff
@@ -24,7 +25,10 @@ struct ApptableArgs {
 #[rustfmt::skip::macros(println, bail)]
-fn apptablecmd(hubris: &mut HubrisArchive, subargs: &[String]) -> Result<()> {
+fn apptablecmd(context: &mut humility::ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
```
I wanted to do this, but the problem is that it doesn't work.

Consider this patch here:

```diff
    //let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
    //let hubris = context.archive.as_mut().unwrap();
    let subargs = context.subargs();
    let hubris = context.hubris();
```

Here's those methods:

```rust
    pub fn subargs(&self) -> &Vec<String> {
        let Subcommand::Other(subargs) = self.cli.cmd.as_ref().unwrap();
        subargs
    }

    pub fn hubris(&mut self) -> &mut HubrisArchive {
        self.archive.as_mut().unwrap()
    }
```

Pretty straightforward. The problem is that this doesn't compile:

```
error[E0502]: cannot borrow `*context` as mutable because it is also borrowed as immutable
  --> cmd\apptable\src\lib.rs:32:18
   |
31 |     let subargs = context.subargs();
   |                   ----------------- immutable borrow occurs here
32 |     let hubris = context.hubris();
   |                  ^^^^^^^^^^^^^^^^ mutable borrow occurs here
33 |
34 |     let subargs = ApptableArgs::try_parse_from(subargs)?;
   |                                                ------- immutable borrow later used here
```

Once you encapsulate them, borrowcheck no longer can say that they're disjoint, because you're not operating on the parts, but instead on the whole thing.

cbiffle replied:
...yeah, I'd leave them as fields for this reason.

steveklabnik commented:
Woo hoo! It's taken a while, but I'm glad to get this finally landed.

Next up: tab complete. 🔜🎊
